File: hmis-engine-api/src/main/java/com/otapp/hmis/engine/masterdata/insurance/application/InsurancePlanService.java
package com.otapp.hmis.engine.masterdata.insurance.application;

import com.otapp.hmis.engine.common.api.PageResponse;
import com.otapp.hmis.engine.common.error.ConflictException;
import com.otapp.hmis.engine.common.error.NotFoundException;
import com.otapp.hmis.engine.masterdata.insurance.application.InsurancePlanDtos.CreateInsurancePlanRequest;
import com.otapp.hmis.engine.masterdata.insurance.application.InsurancePlanDtos.InsurancePlanDto;
import com.otapp.hmis.engine.masterdata.insurance.application.InsurancePlanDtos.UpdateInsurancePlanRequest;
import com.otapp.hmis.engine.masterdata.insurance.domain.InsurancePlan;
import com.otapp.hmis.engine.masterdata.insurance.domain.InsurancePlanRepository;
import com.otapp.hmis.engine.masterdata.insurance.domain.InsuranceProvider;
import com.otapp.hmis.engine.masterdata.insurance.domain.InsuranceProviderRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class InsurancePlanService {

    private final InsurancePlanRepository planRepository;
    private final InsuranceProviderRepository providerRepository;

    @Transactional
    public InsurancePlanDto create(CreateInsurancePlanRequest request) {
        String code = request.code().trim().toUpperCase();
        if (planRepository.existsByCode(code)) {
            throw new ConflictException("Plan code already exists: " + code);
        }
        InsuranceProvider provider = providerRepository.findByUid(request.providerUid())
                .orElseThrow(() -> new NotFoundException("Insurance provider not found: " + request.providerUid()));

        InsurancePlan plan = new InsurancePlan(code, request.name().trim(),
                provider.getUid(), request.description());
        plan.setCoversConsultation(request.coversConsultation());
        plan.setCoversLab(request.coversLab());
        plan.setCoversRadiology(request.coversRadiology());
        plan.setCoversProcedure(request.coversProcedure());
        plan.setCoversMedicine(request.coversMedicine());
        plan.setCoversAdmission(request.coversAdmission());
        planRepository.save(plan);

        return toDto(plan, provider.getName());
    }

    @Transactional
    public InsurancePlanDto update(String uid, UpdateInsurancePlanRequest request) {
        InsurancePlan plan = loadOrThrow(uid);
        plan.setName(request.name().trim());
        plan.setCoversConsultation(request.coversConsultation());
        plan.setCoversLab(request.coversLab());
        plan.setCoversRadiology(request.coversRadiology());
        plan.setCoversProcedure(request.coversProcedure());
        plan.setCoversMedicine(request.coversMedicine());
        plan.setCoversAdmission(request.coversAdmission());
        plan.setDescription(request.description());
        return toDto(plan);
    }

    @Transactional
    public InsurancePlanDto setActive(String uid, boolean active) {
        InsurancePlan plan = loadOrThrow(uid);
        if (active) plan.activate(); else plan.deactivate();
        return toDto(plan);
    }

    @Transactional
    public void delete(String uid) { planRepository.delete(loadOrThrow(uid)); }

    @Transactional(readOnly = true)
    public InsurancePlanDto findByUid(String uid) { return toDto(loadOrThrow(uid)); }

    @Transactional(readOnly = true)
    public PageResponse<InsurancePlanDto> search(String query, Boolean active, String providerUid, Pageable pageable) {
        return PageResponse.from(
                planRepository.search(query == null ? null : query.trim(), active, providerUid, pageable)
                        .map(this::toDto));
    }

    private InsurancePlan loadOrThrow(String uid) {
        return planRepository.findByUid(uid).orElseThrow(() -> new NotFoundException("Insurance plan not found: " + uid));
    }

    private InsurancePlanDto toDto(InsurancePlan p) {
        String providerName = providerRepository.findByUid(p.getProviderUid())
                .map(InsuranceProvider::getName)
                .orElse(null);
        return toDto(p, providerName);
    }

    private static InsurancePlanDto toDto(InsurancePlan p, String providerName) {
        return new InsurancePlanDto(
                p.getUid(), p.getCode(), p.getName(),
                p.getProviderUid(), providerName,
                p.isCoversConsultation(), p.isCoversLab(), p.isCoversRadiology(),
                p.isCoversProcedure(), p.isCoversMedicine(), p.isCoversAdmission(),
                p.getDescription(), p.isActive(), p.getCreatedAt(), p.getUpdatedAt());
    }
}

File: hmis-engine-web/src/app/features/masterdata/insurance-plans/insurance-plan.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';

import { environment } from '../../../../environments/environment';
import { PageResponse } from '../../../core/http/page.types';
import {
  CreateInsurancePlanRequest, InsurancePlan, InsurancePlanSearchParams, UpdateInsurancePlanRequest
} from './insurance-plan.types';

@Injectable({ providedIn: 'root' })
export class InsurancePlanService {
  private readonly http = inject(HttpClient);
  private readonly base = `${environment.apiUrl}/masterdata/insurance-plans`;

  search(params: InsurancePlanSearchParams = {}): Observable<PageResponse<InsurancePlan>> {
    let p = new HttpParams();
    if (params.query) p = p.set('query', params.query);
    if (params.active !== undefined) p = p.set('active', String(params.active));
    if (params.providerUid) p = p.set('providerUid', params.providerUid);
    if (params.page !== undefined) p = p.set('page', String(params.page));
    if (params.size !== undefined) p = p.set('size', String(params.size));
    if (params.sort) p = p.set('sort', params.sort);
    return this.http.get<PageResponse<InsurancePlan>>(this.base, { params: p });
  }

  findByUid(uid: string): Observable<InsurancePlan> { return this.http.get<InsurancePlan>(`${this.base}/${uid}`); }
  create(req: CreateInsurancePlanRequest): Observable<InsurancePlan> { return this.http.post<InsurancePlan>(this.base, req); }
  update(uid: string, req: UpdateInsurancePlanRequest): Observable<InsurancePlan> { return this.http.put<InsurancePlan>(`${this.base}/${uid}`, req); }
  setActive(uid: string, active: boolean): Observable<InsurancePlan> { return this.http.put<InsurancePlan>(`${this.base}/${uid}/active`, { active }); }
  delete(uid: string): Observable<void> { return this.http.delete<void>(`${this.base}/${uid}`); }
}
